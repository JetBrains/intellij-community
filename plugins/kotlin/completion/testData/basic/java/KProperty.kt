// FIR_COMPARISON
// FIR_IDENTICAL
/* Number of result in completion should be fixed after removing duplicates */
package first

fun firstFun() {
  val a = KProp<caret>
}

// INVOCATION_COUNT: 1
// EXIST: { lookupString:"KProperty", itemText:"KProperty", tailText:"<V> (kotlin.reflect)", icon: "org/jetbrains/kotlin/idea/icons/interfaceKotlin.svg"}
