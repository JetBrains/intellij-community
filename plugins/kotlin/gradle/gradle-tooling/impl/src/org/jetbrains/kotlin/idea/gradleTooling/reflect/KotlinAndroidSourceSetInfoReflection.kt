// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.gradleTooling.reflect

fun KotlinAndroidSourceSetInfoReflection(instance: Any): KotlinAndroidSourceSetInfoReflection {
    return KotlinAndroidSourceSetInfoReflectionImpl(instance)
}

interface KotlinAndroidSourceSetInfoReflection {
    val kotlinSourceSetName: String?
    val androidSourceSetName: String?
    val androidVariantNames: Set<String>?
}

private class KotlinAndroidSourceSetInfoReflectionImpl(
    private val instance: Any
) : KotlinAndroidSourceSetInfoReflection {

    override val kotlinSourceSetName: String? by lazy {
        instance.callReflectiveGetter<String>("getKotlinSourceSetName", logger)
    }

    override val androidSourceSetName: String? by lazy {
        instance.callReflectiveGetter<String>("getAndroidSourceSetName", logger)
    }

    override val androidVariantNames: Set<String>? by lazy {
        instance.callReflectiveGetter<Iterable<*>>("getAndroidVariantNames", logger)?.mapNotNull { it as? String }?.toSet()
    }

    companion object {
        val logger = ReflectionLogger(KotlinAndroidSourceSetInfoReflection::class.java)
    }
}