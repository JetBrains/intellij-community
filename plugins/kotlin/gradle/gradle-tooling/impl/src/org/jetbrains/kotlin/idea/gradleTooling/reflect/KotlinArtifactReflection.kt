// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.gradleTooling.reflect

import org.gradle.api.Task
import org.gradle.api.provider.Provider
import org.gradle.api.tasks.Exec
import org.jetbrains.kotlin.idea.gradleTooling.loadClassOrNull
import java.io.File

fun KonanArtifactReflection(konanArtifact: Any): KonanArtifactReflection = KonanArtifactReflectionImpl(konanArtifact)

interface KonanArtifactReflection {
    val executableName: String?
    val compilationTargetName: String?
    val outputKindType: String?
    val konanTargetName: String?
    val outputFile: File?
    val linkTaskPath: String?
    val runTask: Exec?
    val isTests: Boolean?
    val freeCompilerArgs: Collection<String>?
    val binaryOptions: Map<String, String>?
}

private class KonanArtifactReflectionImpl(private val instance: Any) : KonanArtifactReflection {
    override val executableName: String? by lazy {
        instance.callReflectiveGetter("getBaseName", logger)
    }
    private val linkTask: Task? by lazy {
        instance.callReflectiveGetter("getLinkTask", logger)
    }
    override val linkTaskPath: String? by lazy {
        linkTask?.path
    }
    override val runTask: Exec? by lazy {
        if (instance.javaClass.classLoader.loadClassOrNull(NATIVE_EXECUTABLE_CLASS)?.isInstance(instance) == true)
            instance.callReflectiveGetter("getRunTask", logger)
        else null
    }

    override val compilationTargetName: String? by lazy {
        val compilation = linkTask?.callReflectiveAnyGetter("getCompilation", logger)?.let {
            when (it) {
                is Provider<*> -> it.get()
                else -> it
            }
        }
        compilation?.callReflectiveAnyGetter("getTarget", logger)
            ?.callReflectiveGetter("getName", logger)
    }
    override val outputKindType: String? by lazy {
        linkTask?.callReflectiveAnyGetter("getOutputKind", logger)
            ?.callReflectiveGetter("name", logger)
    }
    override val konanTargetName: String? by lazy {
        linkTask?.callReflectiveGetter("getTarget", logger)
    }
    override val outputFile: File? by lazy {
        when (val outputFile = instance.callReflective("getOutputFile", parameters(), returnType<Any>(), logger)) {
            is Provider<*> -> outputFile.orNull as? File
            is File -> outputFile
            else -> null
        }
    }
    override val isTests: Boolean? by lazy {
        linkTask?.callReflectiveGetter("getProcessTests", logger)
    }

    override val freeCompilerArgs: Collection<String>? by lazy {
        linkTask?.callReflectiveAnyGetter("getKotlinOptions", logger)
            ?.callReflectiveGetter("getFreeCompilerArgs", logger)
    }

    override val binaryOptions: Map<String, String>? by lazy {
        linkTask?.callReflectiveGetter("getBinaryOptions", logger)
    }

    companion object {
        private val logger: ReflectionLogger = ReflectionLogger(KonanArtifactReflection::class.java)
        private const val NATIVE_EXECUTABLE_CLASS = "org.jetbrains.kotlin.gradle.plugin.mpp.Executable"
    }
}
