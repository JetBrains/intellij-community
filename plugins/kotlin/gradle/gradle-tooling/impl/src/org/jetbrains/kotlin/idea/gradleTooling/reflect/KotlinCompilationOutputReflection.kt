// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.

package org.jetbrains.kotlin.idea.gradleTooling.reflect

import java.io.File

fun KotlinCompilationOutputReflection(compilationOutputs: Any): KotlinCompilationOutputReflection =
    KotlinCompilationOutputReflectionImpl(compilationOutputs)

interface KotlinCompilationOutputReflection {
    val classesDirs: Iterable<File>?
    val resourcesDir: File?
}

private class KotlinCompilationOutputReflectionImpl(private val instance: Any) : KotlinCompilationOutputReflection {
    override val classesDirs: Iterable<File>? by lazy {
        instance.callReflectiveGetter("getClassesDirs", logger)
    }
    override val resourcesDir: File? by lazy {
        instance.callReflectiveGetter("getResourcesDir", logger)
    }

    companion object {
        val logger = ReflectionLogger(KotlinCompilationOutputReflection::class.java)
    }
}
