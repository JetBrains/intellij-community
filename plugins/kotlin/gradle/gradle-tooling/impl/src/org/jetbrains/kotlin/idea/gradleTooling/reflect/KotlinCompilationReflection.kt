// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.gradleTooling.reflect

import org.gradle.api.Named
import org.jetbrains.kotlin.idea.gradleTooling.getMethodOrNull

fun KotlinCompilationReflection(kotlinCompilation: Any): KotlinCompilationReflection =
    KotlinCompilationReflectionImpl(kotlinCompilation)

interface KotlinCompilationReflection {
    val target: KotlinTargetReflection?
    val compilationName: String
    val gradleCompilation: Named
    val sourceSets: Collection<Named>? // Source Sets that directly added to compilation
    val allSourceSets: Collection<Named>? // this.sourceSets + their transitive closure through dependsOn relation
    val compilationOutput: KotlinCompilationOutputReflection?
    val konanTargetName: String?
    val compileKotlinTaskName: String?
    val associateCompilations: Iterable<KotlinCompilationReflection>
}

private class KotlinCompilationReflectionImpl(private val instance: Any) : KotlinCompilationReflection {

    override val target: KotlinTargetReflection? by lazy {
        KotlinTargetReflection(instance.callReflectiveAnyGetter("getTarget", logger) ?: return@lazy null)
    }

    override val gradleCompilation: Named
        get() = instance as Named

    override val compilationName: String by lazy {
        gradleCompilation.name
    }
    override val sourceSets: Collection<Named>? by lazy {
        instance.callReflectiveGetter("getKotlinSourceSets", logger)
    }

    override val allSourceSets: Collection<Named>? by lazy {
        instance.callReflectiveGetter("getAllKotlinSourceSets", logger)
    }

    override val compilationOutput: KotlinCompilationOutputReflection? by lazy {
        instance.callReflectiveAnyGetter("getOutput", logger)?.let { gradleOutput -> KotlinCompilationOutputReflection(gradleOutput) }
    }

    // Get konanTarget (for native compilations only).
    override val konanTargetName: String? by lazy {
        if (instance.javaClass.getMethodOrNull("getKonanTarget") == null) null
        else instance.callReflectiveAnyGetter("getKonanTarget", logger)
            ?.callReflectiveGetter("getName", logger)
    }
    override val compileKotlinTaskName: String? by lazy {
        instance.callReflectiveGetter("getCompileKotlinTaskName", logger)
    }

    override val associateCompilations: Iterable<KotlinCompilationReflection> by lazy {
        instance.callReflectiveGetter<List<*>>("getAssociateWith", logger).orEmpty()
            .filterNotNull()
            .map { compilation -> KotlinCompilationReflection(compilation) }
    }

    companion object {
        private val logger: ReflectionLogger = ReflectionLogger(KotlinCompilationReflection::class.java)
    }
}
