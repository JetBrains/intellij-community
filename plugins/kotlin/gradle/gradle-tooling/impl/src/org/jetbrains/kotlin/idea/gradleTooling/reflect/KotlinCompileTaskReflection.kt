// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.gradleTooling.reflect

import org.gradle.api.Task

fun KotlinCompileTaskReflection(task: Task): KotlinCompileTaskReflection {
    return KotlinCompileTaskReflectionImpl(task)
}

interface KotlinCompileTaskReflection {
    val task: Task
    val serializedCompilerArguments: List<String>
}

private class KotlinCompileTaskReflectionImpl(override val task: Task) : KotlinCompileTaskReflection {
    override val serializedCompilerArguments: List<String>
        get() = runCatching {
            task.callReflectiveGetter<List<String>>("getSerializedCompilerArguments", logger)
        }.recoverCatching {
            logger.logIssue("Failed calling 'getSerializedCompilerArguments'", it)
            task.callReflectiveGetter<List<String>>("getSerializedCompilerArgumentsIgnoreClasspathIssues", logger)
        }.onFailure {
            logger.logIssue("Failed getting 'serializedCompilerArguments' from '${task.path}'")
        }.getOrNull().orEmpty()


    companion object {
        val logger = ReflectionLogger(KotlinCompileTaskReflection::class.java)
    }
}
