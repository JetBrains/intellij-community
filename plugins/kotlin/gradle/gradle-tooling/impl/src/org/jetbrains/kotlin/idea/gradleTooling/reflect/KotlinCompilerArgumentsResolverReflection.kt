// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.gradleTooling.reflect

import org.gradle.api.Project
import org.jetbrains.kotlin.idea.gradleTooling.loadClassOrNull

fun KotlinCompilerArgumentsResolverReflection(project: Project, classLoader: ClassLoader): KotlinCompilerArgumentsResolverReflection? {
    val clazz = classLoader.loadClassOrNull("org.jetbrains.kotlin.gradle.plugin.ide.IdeCompilerArgumentsResolver") ?: return null

    val instance = Static(clazz).callReflective(
        "instance", parameters(parameter(project)), returnType<Any>(), KotlinMultiplatformImportReflectionImpl.logger
    ) ?: return null

    return KotlinCompilerArgumentsResolverReflectionImpl(instance)
}

interface KotlinCompilerArgumentsResolverReflection {
    fun resolveCompilerArguments(owner: Any): List<String>?
}

private class KotlinCompilerArgumentsResolverReflectionImpl(val instance: Any) : KotlinCompilerArgumentsResolverReflection {
    override fun resolveCompilerArguments(owner: Any): List<String>? {
        return instance.callReflective(
            "resolveCompilerArguments", parameters(parameter<Any>(owner)), returnType<List<String>?>(),
            KotlinMultiplatformImportReflectionImpl.logger
        )
    }
}
