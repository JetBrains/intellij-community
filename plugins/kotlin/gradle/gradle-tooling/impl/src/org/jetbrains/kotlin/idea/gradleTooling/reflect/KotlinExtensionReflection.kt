// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.gradleTooling.reflect

import org.gradle.api.Named
import org.gradle.api.Project

class KotlinExtensionReflection(
    val project: Project, val kotlinExtension: Any
) {
    val kotlinGradlePluginVersion: String? by lazy {
        Static(KOTLIN_PLUGIN_WRAPPER_FILE_CLASS_NAME, kotlinExtension.javaClass.classLoader, logger)
            ?.callReflective("getKotlinPluginVersion", parameters(parameter(project)), returnType<String>(), logger)
    }

    val targets: List<KotlinTargetReflection> by lazy {
        kotlinExtension.callReflectiveGetter<Iterable<*>>("getTargets", logger)?.filterNotNull()
            ?.map { KotlinTargetReflection(it) }.orEmpty()
    }

    val sourceSets: List<KotlinSourceSetReflection> by lazy {
        kotlinExtension.callReflectiveGetter<Iterable<*>>("getSourceSets", logger)
            ?.filterNotNull().orEmpty().mapNotNull { sourceSet ->
                if (sourceSet !is Named) {
                    logger.logIssue("KotlinSourceSet $sourceSet does not implement 'Named'")
                    return@mapNotNull null
                }
                sourceSet
            }
            .map { sourceSet -> KotlinSourceSetReflection(sourceSet) }
    }

    val sourceSetsByName: Map<String, KotlinSourceSetReflection> by lazy {
        sourceSets.associateBy { it.name }
    }

    companion object {
        private val logger: ReflectionLogger = ReflectionLogger(KotlinExtensionReflection::class.java)
        private const val KOTLIN_PLUGIN_WRAPPER_FILE_CLASS_NAME = "org.jetbrains.kotlin.gradle.plugin.KotlinPluginWrapperKt"
    }
}