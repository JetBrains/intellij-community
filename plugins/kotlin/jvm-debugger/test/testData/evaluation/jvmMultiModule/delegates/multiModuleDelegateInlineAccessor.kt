// MODULE: jvm-lib
// FILE: myVeryOwnInlineDelegates.kt

package myVeryOwnInlineDelegates

import kotlin.reflect.KProperty

interface MyDelegate<T> {
    val value: T
}

class MyDelegateImpl<T>(override val value: T) : MyDelegate<T>

public inline operator fun <T> MyDelegate<T>.getValue(thisRef: Any?, property: KProperty<*>): T = value

fun <T> myDelegate(init: () -> T) = MyDelegateImpl(init())

// MODULE: jvm-app(jvm-lib)
// FILE: multiModuleDelegateInlineAccessor.kt

import myVeryOwnInlineDelegates.myDelegate
import myVeryOwnInlineDelegates.getValue

class Delegates {
    var text: String = ""

    val myMapping by myDelegate {
        mapOf(
            "Hello" to 1,
            "World" to 2
        )
    }

    fun mapped(i: Int) = myMapping.entries.first { it.value == i }.key

    fun foo(i: Int) {
        text = mapped(i)
        // EXPRESSION: text
        // RESULT: "World": Ljava/lang/String;
        //Breakpoint!
        println(text)
    }
}

fun main() {
    Delegates().foo(2)
}
