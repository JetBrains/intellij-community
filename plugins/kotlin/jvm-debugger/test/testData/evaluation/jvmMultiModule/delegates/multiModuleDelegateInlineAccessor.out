LineBreakpoint created at multiModuleDelegateInlineAccessor.kt:24
Run Java
Connected to the target VM
multiModuleDelegateInlineAccessor.kt:24
Compile bytecode for text
Disconnected from the target VM

Process finished with exit code 0
